tests/ui/fn/fn_def_coercion.rs

```diff
@@ -0,0 +1,57 @@
+//! Test that coercing between function items of the same function,
+//! but with different generic args works.
```

Comment:
this test has a lot of region errors 🤔 

I guess you want to test that coercions succeed and we only get the issue during mir typeck because the args of the function are invariant? Please explain the rationale here

Reply:
I just looked through tests that were relevant to these coercions, and realized the code paths were undertested, so I added a bunch of tests that hit this code path without using opaques